File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week7/MoveX.java
/*
MoveX.java
Billy Cussen
18/11/2020
*/

public class MoveX{

    private int spaceTrigger = 1;

    public static void main(String[]args){

        MoveX myMoveX = new MoveX();

        long startTime = System.nanoTime();
        System.out.println("Move X: "+moveX("HxexlXlxoX"));
        long endTime = System.nanoTime();
        System.out.println("Looped Approach took "+(endTime-startTime)+" nanoseconds");
        
        startTime = System.nanoTime();
        System.out.println("Move X Recursion: "+myMoveX.moveXRecursion("HxexlXlxoX"));
        endTime = System.nanoTime();
        System.out.println("Recursive Approach took "+(endTime-startTime)+" nanoseconds");
    }

    public static String moveX(String word){
        StringBuffer sb = new StringBuffer();
        String x = "";
        for(int i = 0; i < word.length(); i++){
            if(word.charAt(i)=='X' || word.charAt(i)=='x'){
                x += word.charAt(i);
            } else{
                sb.append(word.charAt(i));
            }
        }
        sb.append(" "+x);
        return sb.toString();
    }

    public String moveXRecursion(String word){
        if(word.length()==0){
            return word;
        }

        if(spaceTrigger==1){
            word+= " ";
            spaceTrigger++;
        }

        if(word.charAt(0)=='X' || word.charAt(0)=='x'){
            return moveXRecursion(word.substring(1,word.length()))+word.charAt(0);
        } else{
            return word.charAt(0) + moveXRecursion(word.substring(1,word.length()));
        }
    }

}

File: Python/Data Structures & Algorithms Module - Python/Week7/MoveX.py
"""
MoveX.py
Billy Cussen
18/11/2020
"""
import timeit

spaceCount = 1

def moveX(word):
    x, newWord = "", ""

    for i in word:
        if i == 'x' or i == 'X':
            x += i
        else:
            newWord += i
    return newWord+" "+x
    
def moveXRecursion(word):
    if len(word) == 0:
        return word

    global spaceCount
    if spaceCount == 1:
        word += " "
        spaceCount+= 1

    if word[0] == 'x' or word[0] == 'X':
        return moveXRecursion(word[1:])+word[0]
    else:
        return word[0] + moveXRecursion(word[1:])

startTime = timeit.default_timer()
print("Move X: "+str(moveX("HxexlXlxox Txhxexrxex!")))
endTime = timeit.default_timer() 
nanoTime = round((endTime-startTime)*1000000000,2)
print("Looped Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer() 
print("Move X Recursion: "+str(moveXRecursion("HxexlXlxox Txhxexrxex!")))
endTime = timeit.default_timer() 
nanoTime = round((endTime-startTime)*1000000000,2)
print("Recursive Approach took "+str(nanoTime)+" nanoseconds")